scalin groups: basic sample (#53)
kedify/examples: samples/scaling-group/README.md

# ScalingGroups

This is an example usage of a kedify `ScalingGroup` feature to enforce a limit on scaling shared among a group of `ScaledObjects`.

> Currently, the feature is guarded by feature flag, to enable this set following env var on a `keda-operator`.

Reviewer: ```suggestion
Currently, the feature is guarded by feature flag. To enable this, set the following env var on a `keda-operator`.
```

```

> $ kubectl --namespace=keda set env deployment/keda-operator KEDIFY_ENABLED_SCALINGGROUPS=true

Reviewer 2: ```suggestion
$ kubectl --namespace=keda set env deployment/keda-operator KEDIFY_SCALINGGROUPS_ENABLED=true
```

Change this if we decide to change this

```

### Basic Group Capacity

There is an example of a `ScalingGroup`, you can deploy it to your cluster
```
$ kubectl apply -f scalinggroup.yaml
```

The controller for scaling groups should populate the status, currently there are no `ScaledObjects` matching the defined selector
so the status should look similar to this:
```
$ kubectl get sg
NAME        CAPACITY   RESIDUALCAPACITY   MEMBERCOUNT
max-group   2          2                  0
```

Provided are two applications with `ScaledObject` that match the group selector, both applications use `kubernetes-workload` scaler

> which makes arbitrary scaling for testing easily possible. Lets deploy first one

Reviewer: ```suggestion
which makes arbitrary scaling for testing easily possible. Let's deploy the first one
```

```
There are now two `Deployments` with one `Pod` each and a single `ScaledObject`
```
$ kubectl get deployments
NAME                 READY   UP-TO-DATE   AVAILABLE   AGE
app-1                1/1     1            1           10s
test-keda-metric-1   1/1     1            1           10s

$ kubectl get scaledobject
NAME    SCALETARGETKIND      SCALETARGETNAME   MIN   MAX   READY   ACTIVE   FALLBACK   PAUSED    TRIGGERS   AUTHENTICATIONS   AGE
app-1   apps/v1.Deployment   app-1             0     10    True    True     False      Unknown                                11s
```
The `Deployment` with name `app-1` is scaled by `app-1` `ScaledObject` which has a metric source number of `Pods` in `test-keda-metric-1`
mock `Deployment`. When `test-keda-metric-1` is scaled to higher replica count, `HPA` will also try to scale `app-1` to the matching replica
count.

The controller for scaling groups will adjust the `ScalingGroup` status
```
$ kubectl get sg
NAME        CAPACITY   RESIDUALCAPACITY   MEMBERCOUNT
max-group   2          1                  1
```

At this point there is no need to cap any metrics because the total sum of replicas across all member `ScaledObjects` is below

> the group capacity. We can try to change that with increasing the underlying metric source for the `ScaledObject`

Reviewer: ```suggestion
the group capacity. We can try to change that by increasing the underlying metric source for the `ScaledObject`
```

...
Events:
  Type    Reason                    Age   From                        Message
  ----    ------                    ----  ----                        -------
  Normal  AddedToScalingGroup       24s   sg-scaledobject-controller  ScaledObject added to ScalingGroup max-group
  Normal  KEDAScalersStarted        24s   keda-operator               Scaler kubernetes-workload is built.
  Normal  KEDAScalersStarted        24s   keda-operator               Started scalers watch
  Normal  ScaledObjectReady         24s   keda-operator               ScaledObject is ready for scaling
  Normal  KEDAScaleTargetActivated  24s   keda-operator               Scaled apps/v1.Deployment default/app-1 from 0 to 1, triggered by kubernetesWorkloadScaler
  Normal  MetricsCapped             17s   sg-metrics-processor        Metrics capped: [s0-workload-default(5 => 2)] due to ScalingGroup max-group
```
There are well known events from `keda-operator` and also two new. One from `sg-scaledobject-controller` regarding `ScaledObject` being
added to a group and second from `sg-metrics-processor` about metric being capped to satisfy group capacity.

### Overprovisioning Recovery

> Now when the group is at max capacity, adding a new `ScaledObject` would result in overprovisioning. There is a built in algorithm to

Reviewer: ```suggestion
Now when the group is at max capacity, adding a new `ScaledObject` would result in overprovisioning. There is a built-in algorithm to
```

gradually and eventually decrease metrics for `ScaledObjects` in the overprovisioned group to try to satisfy the group capacity.

> Lets create second autoscaled application, it is a mirror of the `app-1`, similar deployments, similar scaling triggers, only called `app-2`.

Reviewer: ```suggestion
Let's create a second autoscaled application, it is a mirror of the `app-1`, similar deployments, similar scaling triggers, only called `app-2`.
```

Reviewer: maybe `s/second/a second/`? not sure
